discovery-provider/src/utils/helpers.py

```diff
@@ -14,6 +14,44 @@
 from src import exceptions
 from . import multihash
 
+def redis_restore(redis, key):
+    logger = logging.getLogger(__name__)
+    try:
+        filename = f"{key}_dump"
+        with open(filename, 'rb') as f:
+            dumped = f.read()
+            redis.restore(key, 0, dumped)
+            logger.info(f"successfully restored redis value for key: {key}")
+            return redis.get(key)
+    except FileNotFoundError as not_found:
+        logger.error(f"could not read redis dump file: {filename}")
+        logger.error(not_found)
+        return None
+    except Exception as e:
+        logger.error(f"could not perform redis restore for key: {key}")
+        logger.error(e)
+        return None
+
+def redis_dump(redis, key):
+    logger = logging.getLogger(__name__)
+    try:
+        dumped = redis.dump(key)
+        filename = f"{key}_dump"
+        with open(filename, 'wb') as f:
+            f.write(dumped)
+            logger.info(f"successfully performed redis dump for key: {key}")
+    except Exception as e:
+        logger.error(f"could not perform redis dump for key: {key}")
+        logger.error(e)
+
+def redis_set_and_dump(redis, key, value):
+    redis.set(key, value)
+    redis_dump(redis, key)
+
+def redis_get_or_restore(redis, key):
```

Comment:
just so i understand, this restores from the previous dump if a key doesn't exist? would this cause any data loss issues if a key that doesn't exist for some reason/was incorrectly formatted was passed in?

Reply:
you got that right! there are cases when the key does not exist e.g. the first time you try to get without having priorly set. biggest consequences of losing the cached data would be the lack of ability to deduplicate some of the data, or that the last few minutes of a node's metrics are lost, but risk of data loss seems low to me as the keys are static strings and there is no formatting foo

Reply:
gotcha. i guess there's no harm in calling it this way because the redis data is backed up/restored on a by key basis. so if the key doesn't exist, no harm in restoring from disk. i was thinking more along the lines of data corruption or data loss but if the value in redis is null, that's not even possible